sycl/include/sycl/ext/intel/experimental/esimd/memory.hpp

```diff
@@ -967,23 +1011,38 @@ lsc_block_store(AccessorTy acc, uint32_t offset,
   lsc_block_store<T, NElts, DS, L1H>(
       __ESIMD_DNS::accessorToPointer<T>(acc, offset), vals, pred);
 #else
-  detail::check_lsc_vector_size<NElts>();
   detail::check_lsc_data_size<T, DS>();
-  detail::check_lsc_cache_hint<detail::lsc_action::store, L1H, L3H>();
+  detail::check_lsc_cache_hint<detail::lsc_action::load, L1H, L3H>();
   constexpr uint16_t _AddressScale = 1;
   constexpr int _ImmOffset = 0;
   constexpr lsc_data_size _DS = detail::finalize_data_size<T, DS>();
-  static_assert(_DS == lsc_data_size::u32 || _DS == lsc_data_size::u64,
-                "Transposed store is supported only for data size u32 or u64");
-  constexpr detail::lsc_vector_size _VS = detail::to_lsc_vector_size<NElts>();
   constexpr detail::lsc_data_order _Transposed =
       detail::lsc_data_order::transpose;
   constexpr int N = 1;
+
   __ESIMD_NS::simd<uint32_t, N> offsets = offset;
   auto si = __ESIMD_NS::get_surface_index(acc);
-  __esimd_lsc_store_bti<T, L1H, L3H, _AddressScale, _ImmOffset, _DS, _VS,
-                        _Transposed, N>(pred.data(), offsets.data(),
-                                        vals.data(), si);
+
+  if constexpr (_DS == lsc_data_size::u32 || _DS == lsc_data_size::u64) {
+    detail::check_lsc_vector_size<NElts>();
+    constexpr detail::lsc_vector_size _VS = detail::to_lsc_vector_size<NElts>();
+    __esimd_lsc_store_bti<T, L1H, L3H, _AddressScale, _ImmOffset, _DS, _VS,
+                          _Transposed, N>(pred.data(), offsets.data(),
+                                          vals.data(), si);
+  } else if constexpr (_DS == lsc_data_size::u16 || _DS == lsc_data_size::u8) {
+    constexpr int NElemsInDword = _DS == lsc_data_size::u16 ? 2 : 4;
+    static_assert(NElts % NElemsInDword == 0,
+                  "Number of elements is not supported by Transposed store");
+    detail::check_lsc_vector_size<NElts / NElemsInDword>();
+    constexpr detail::lsc_vector_size _VS =
+        detail::to_lsc_vector_size<NElts / NElemsInDword>();
+    __ESIMD_NS::simd<uint32_t, NElts / NElemsInDword> tmp =
+        vals.template bit_cast_view<uint32_t>();
+
+    __esimd_lsc_store_bti<uint32_t, L1H, L3H, _AddressScale, _ImmOffset,
+                          lsc_data_size::u32, _VS, _Transposed, N>(
+        pred.data(), offsets.data(), tmp.data(), si);
+  }
```

Comment:
Thank you for combining U8 and U16 cases. After that was done, it became more obvious that major part of code in 'if-then' and 'if-else' branches is similar.
It is not the requirement, but this code could be reduced even more:
```suggestion
  constexpr int SmallIntFactor =
      (_DS == lsc_data_size::u16) ? 2 : (_DS == lsc_data_size::u8 ? 4 : 1);
  static_assert(NElts % SmallIntFactor == 0,
                "Number of elements is not supported by Transposed store");
  detail::check_lsc_vector_size<NElts / SmallIntFactor>();
  constexpr auto _VS = detail::to_lsc_vector_size<NElts / SmallIntFactor>();
  if constexpr (SmallIntFactor > 1) {
    auto Tmp = vals.template bit_cast_view<uint32_t>();
    __esimd_lsc_store_bti<uint32_t, L1H, L3H, _AddressScale, _ImmOffset,
                          lsc_data_size::u32, _VS, _Transposed, N>(
        pred.data(), offsets.data(), Tmp.data(), si);
  } else {
    __esimd_lsc_store_bti<T, L1H, L3H, _AddressScale, _ImmOffset, _DS, _VS,
                          _Transposed, N>(pred.data(), offsets.data(),
                                          vals.data(), si);
  }
```